src/frontend/flight/sync/writer.rs

```diff
@@ -158,36 +161,54 @@ impl SeafowlDataSyncWriter {
     pub fn enqueue_sync(
         &mut self,
         log_store: Arc<dyn LogStore>,
-        sequence_number: SequenceNumber,
+        sequence_number: Option<SequenceNumber>,
         origin: Origin,
         sync_schema: SyncSchema,
-        last: bool,
         batches: Vec<RecordBatch>,
-    ) -> Result<()> {
+    ) -> SyncResult<()> {
         let url = log_store.root_uri();
 
         // Upsert a sequence entry for this origin and sequence number
-        let sequence = DataSyncSequence {
-            last,
-            locs: HashSet::from([url.clone()]),
-        };
-        self
-            .seqs
-            .entry(origin.clone())
-            .and_modify(|origin_seqs| {
-                origin_seqs.entry(sequence_number).and_modify(|seq| {if !seq.locs.contains(&url) {
-                    debug!("Adding {url} as sync destination for {origin}, {sequence_number}");
-                    seq.locs.insert(url.clone());
-                }
+        let tx_id = if let Some((tx_id, tx)) = self.txs.last_mut()
+            && tx.sequence.is_none()
+        {
+            if origin != tx.origin {
+                return Err(SyncError::InvalidMessage {
+                    reason: format!(
+                        "Transaction from a new origin ({origin}) started without finishing the transaction from the previous one ({})",
+                        tx.origin
+                    )
+                });
```

Comment:
This can happen during failover though? @eatonphil

Reply:
Yes I think I commented about this in tests Marko mentioned me on. You should not panic. While it's an invalid scenario it is certainly a possible scenario. You should not crash or become unrecoverable from this.

Reply:
Another example is not even failover but just if the writer process itself crashes midway through. We may not send the full transaction and then may send it again (from the same origin).

Reply:
Note that the above merely returns an error to the caller, and doesn't panic/crash the process. Then presumably the other side can re-adjust it's position in the message queue and try again?